shell/platform/embedder/embedder.cc

```diff
@@ -2326,6 +2326,7 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
     pointer_data.pan_delta_y = 0.0;
     pointer_data.scale = SAFE_ACCESS(current, scale, 0.0);
     pointer_data.rotation = SAFE_ACCESS(current, rotation, 0.0);
+    pointer_data.view_id = SAFE_ACCESS(current, view_id, 0);
```

Comment:
Can we use `kFlutterImplicitViewId` here? (FYI in the future the embedder will control the implicit view ID, so this will need to updated later)

```suggestion
    pointer_data.view_id = SAFE_ACCESS(current, view_id, kFlutterImplicitViewId);
```

Reply:
I'm kind of hesitant here. The benefit of using 0 as the default value is that we can flush the struct to be all 0s and be sure that it doesn't break even when new fields are added. But I also agree that the default value should likely be the implicit view. In this way, we should probably NOT allow embedders to customize the implicit view ID to be non-0.

Reply:
> The benefit of using 0 as the default value is that we can flush the struct to be all 0s and be sure that it doesn't break even when new fields are added

I don't follow. The embedder API - including the new multi-view APIs - is designed for forwards/backwards ABI compatibility. The `struct_size` member lets us determine how to handle the `view_id` without needing to fallback to `0`.

> In this way, we should probably NOT allow embedders to customize the implicit view ID to be non-0.

I don't understand this either, can you expand on this? The embedder API design doc had a section that allows the embedder to customize the implicit view ID: https://docs.google.com/document/d/1kegGJ235y1778-vzyyv90MgSHmYQcROTupMuLtEDBq0/edit#heading=h.1enoc2wkpbna

Reply:
> The struct_size member lets us determine how to handle the view_id without needing to fallback to 0.

I'm talking about creating such a struct. For example, the following line from the GLFW example (somewhat simplified):

```cpp
void GLFWcursorPositionCallbackAtPhase(GLFWwindow* window,
                                       FlutterPointerPhase phase,
                                       double x,
                                       double y) {
  FlutterPointerEvent event = {};
  event.struct_size = sizeof(event);
  event.phase = phase;
  event.x = x * g_pixelRatio;
  event.y = y * g_pixelRatio;
  // event.view_id = ??? // Does not exist in old apps
  FlutterEngineSendPointerEvent(&event, 1);
}
```

This code is actually problematic. Old apps do not assign `event.view_id`. Therefore, once Flutter adds this new field `view_id` and this custom embedder is **recompiled** without editing, every pointer event Flutter received will have an uninitialized `view_id`, which will be a random number, because `FlutterPointerEvent` is a C struct and do not have default values. this PR.

The `struct_size` guard only works for linking existing compiled libraries.

Therefore, truly backward-compatible code needs to flush all bytes of C structs to 0.

```cpp
void GLFWcursorPositionCallbackAtPhase(GLFWwindow* window,
                                       FlutterPointerPhase phase,
                                       double x,
                                       double y) {
  FlutterPointerEvent event = {};
  memset(this, 0, sizeof(FlutterPointerEvent));  // Add this line
  event.struct_size = sizeof(event);
  event.phase = phase;
  event.x = x * g_pixelRatio;
  event.y = y * g_pixelRatio;
  FlutterEngineSendPointerEvent(&event, 1);
}
```

However, this is equivalent to setting `event.view_id = 0;`.

Reply:
There's two cases that are handled differently:

1. **Embedders in the engine repo** - These automatically use the latest embedder API whenever the API changes.
2. **Embedders outside the engine repo** - These must update their embedder API manually:
    1. [Generate a `flutter_embedder.h` file](https://medium.com/flutter/flutter-on-raspberry-pi-mostly-from-scratch-2824c5e7dcb1)
    2. Copy the `flutter_embedder.h` file into the embedder's project
    3. Update the embedder's logic to use the latest embedder API
    4. Recompile the embedder

See [Chinmay's minimal embedder example](https://github.com/chinmaygarde/flutter_from_scratch/blob/master/README.md).

The `struct_size` guard allows us to detect whether an "outside" embedder has updated their `flutter_embedder.h` file. If an embedder outside the engine repo sends a default value for the view ID, it means the embedder updated their `flutter_embedder.h` file without updating their embedder logic - this would be an error by the "outside" embedder.

Reply:
If the convention is to copy a separate `flutter_embedder.h` then that's fine. But is it mentioned anywhere? Chinmay's example is just his own way. We might want to ask how Flutter defines compatibility here.

Reply:
Here's where "outside" embedders put their copy of `flutter_embedder.h`:

* Sony: https://github.com/sony/flutter-embedded-linux/blob/master/src/flutter/shell/platform/embedder/embedder.h
* Samsung Tizen: https://github.com/flutter-tizen/embedder/blob/master/flutter/shell/platform/embedder/embedder.h
* Toyota: https://github.com/toyota-connected/ivi-homescreen/blob/main/third_party/flutter/shell/platform/embedder/embedder.h
* Go Flutter: https://github.com/go-flutter-desktop/go-flutter/blob/master/embedder/embedder.h
* Raspberry pi: https://github.com/ardera/flutter-pi/blob/master/third_party/flutter_embedder_header/include/flutter_embedder.h

Reply:
I see. In that case we don't have to be limited to use 0 as the default value. And we don't have to use `disable_implicit_view`, since we can make it `enable_implicit_view` and make the default value (by `struct_size`) true.

Reply:
> At least I don't think it's not worth the "customizability" that I doubt if many people would use.

It feels a bit late to change this aspect of the design. This design has been out for over a month and has been reviewed by several stakeholders. We can still pivot if this is truly not worth the effort, but I'm not convinced that's the case.

Reply:
> And we don't have to use `disable_implicit_view`, since we can make it `enable_implicit_view` and make the default value (by `struct_size`) true.

We discussed this with Chinmay but decided to stick to `disable_implicit_view` regardless. The poor name does minimize the impact to a custom embedder that is updated incorrectly (assuming they zero initializes)

Reply:
> It feels a bit late to change this aspect of the design. This design has been out for over a month and has been reviewed by several stakeholders. We can still pivot if this is truly not worth the effort, I'm not convinced that's the case (and it takes effort to review design updates!)

It's never too late to make improvement as long as it has not been submitted into `embedder.h`. Not to mention the fact that the implicit view ID and its customization are probably among last features of this project to add. If we worry about the minimizing the impact to a custom embedder that is updated incorrectly  for `disable_implicit_view`, then we probably want to make it similar for the view ID.

For this particular change though, I'll change it to `kFlutterImplicitViewId` since I agree that the default value should be the implicit view ID, no matter what that is.